src/Functions/vectorFunctions.cpp

```diff
@@ -0,0 +1,1067 @@
+                throw;
+            }
+        }
+
+        return res_type;
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        const auto * cur_tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+        const auto & cur_types = cur_tuple->getElements();
+        auto cur_elements = getTupleElements(*arguments[0].column);
+
+        size_t tuple_size = cur_elements.size();
+        if (tuple_size == 0)
+            return DataTypeUInt8().createColumnConstWithDefaultValue(input_rows_count);
+
+        auto abs = FunctionFactory::instance().get("abs", context);
+        auto max = FunctionFactory::instance().get("max2", context);
+        ColumnWithTypeAndName res;
+        for (size_t i = 0; i < tuple_size; ++i)
+        {
+            ColumnWithTypeAndName cur{cur_elements[i], cur_types[i], {}};
+            auto elem_abs = abs->build(ColumnsWithTypeAndName{cur});
+
+            ColumnWithTypeAndName column;
+            column.type = elem_abs->getResultType();
+            column.column = elem_abs->execute({cur}, column.type, input_rows_count);
+
+            if (i == 0)
+            {
+                res = std::move(column);
+            }
+            else
+            {
+                auto max_elem = max->build({res, column});
+                auto res_type = max_elem->getResultType();
+                res.column = max_elem->execute({res, column}, res_type, input_rows_count);
+                res.type = res_type;
+            }
+        }
+
+        return res.column;
+    }
+};
+using FunctionLinfNorm = FunctionLNorm<Linf_LABEL>;
+
+template <>
+class FunctionLNorm<Lp_LABEL> : public TupleIFunction
+{
+public:
+    static constexpr auto name = "LpNorm";
+
+    explicit FunctionLNorm(ContextPtr context_) : TupleIFunction(context_) {}
+    static FunctionPtr create(ContextPtr context_) { return std::make_shared<FunctionLNorm>(context_); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        const auto * cur_tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+
+        if (!cur_tuple)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Argument 0 of function {} should be tuples, got {}",
+                            getName(), arguments[0].type->getName());
+
+        const auto & cur_types = cur_tuple->getElements();
+
+        Columns cur_elements;
+        if (arguments[0].column)
+            cur_elements = getTupleElements(*arguments[0].column);
+
+        size_t tuple_size = cur_types.size();
+        if (tuple_size == 0)
+            return std::make_shared<DataTypeUInt8>();
+
+        const auto & p_column = arguments[1];
```

Comment:
Do you suggest extracting `p` from the column and the just make a column with `1/p` with type `Float64` in the end? If so, it will inevitably convert the argument to `Float64`, although eventually `pow` is guaranteed to return `Float64` so this is not a problem. I can do it.
Maybe we can just return `Float64` in `getReturnTypeImpl` as `pow` result type is always `Float64`?

Reply:
Simplified in 53d7649c33bb43410b4e1ad52b820da4eebb327f.